File: content/tracks/dsa/problems/lc-0042-trapping-rain-water/solution.py
from typing import List


class Solution:
    def trap(self, height: List[int]) -> int:
        # Two pointers: the side with the lower wall so far is bounded by that wall, so its water
        # can be counted and the pointer moved inward.
        left, right = 0, len(height) - 1
        left_max = right_max = 0
        water = 0
        while left < right:
            if height[left] < height[right]:
                left_max = max(left_max, height[left])
                water += left_max - height[left]
                left += 1
            else:
                right_max = max(right_max, height[right])
                water += right_max - height[right]
                right -= 1
        return water

File: content/tracks/dsa/problems/lc-0042-trapping-rain-water/solution.go
package main

func trap(height []int) int {
	// Two pointers: the side with the lower wall so far is bounded by that wall, so its water
	// can be counted and the pointer moved inward.
	left, right := 0, len(height)-1
	leftMax, rightMax := 0, 0
	water := 0
	for left < right {
		if height[left] < height[right] {
			leftMax = max(leftMax, height[left])
			water += leftMax - height[left]
			left++
		} else {
			rightMax = max(rightMax, height[right])
			water += rightMax - height[right]
			right--
		}
	}
	return water
}
